chainer/link.py

```diff
@@ -139,25 +149,83 @@ def xp(self):
         """
         return numpy if self._cpu else cuda.cupy
 
+    @property
+    def within_init_scope(self):
+        """True if the current code is inside of an initialization scope.
+
+        See :meth:`init_scope` for the details of the initialization scope.
+
+        """
+        return getattr(self, '_within_init_scope', False)
+
+    @contextlib.contextmanager
+    def init_scope(self):
+        """Creates an initialization scope.
+
+        This method returns a context manager object that enables registration
+        of parameters (and links for :class:`~chainer.Chain`) by substitution.
+        A :class:`~chainer.Parameter` object can be automatically registered
+        by substituting it to an attribute under this context manager.
+
+        .. admonition:: Example
+
+           In most cases, the parameter registration is done in the
+           initializer method. Using the ``init_scope`` method, we can
+           simply substitute a :class:`~chainer.Parameter` object to register
+           it to the link.
+
+           class MyLink(chainer.Link):
+               def __init__(self):
+                   super().__init__()
+                   with self.init_scope():
+                       self.W = chainer.Parameter(0, (10, 5))
+                       self.b = chainer.Parameter(0, (5,))
+
+        """
+        old_flag = self.within_init_scope
+        self._within_init_scope = True
+        try:
+            yield
```

Comment:
Because otherwise the code after `yield` will not be executed when an exception is raised in the `with` block.